src/WebApiContrib.Formatting.MsgPack/MessagePackMediaTypeFormatter.cs

```diff
@@ -70,23 +74,35 @@ public override object ReadFromStream(Type type, System.IO.Stream readStream, Sy
                 result = GetDefaultValueForType(type);
             }
 
-            return result;
+            return Task.FromResult(result);
         }
 
-        public override void WriteToStream(Type type, object value, System.IO.Stream writeStream, System.Net.Http.HttpContent content)
+        public override Task WriteToStreamAsync(Type type, object value, Stream stream, HttpContent content, TransportContext transportContext)
         {
             if (type == null)
-                throw new ArgumentNullException("type is null");
+                throw new ArgumentNullException("type");
 
-            if (writeStream == null)
-                throw new ArgumentNullException("writeStream is null");
+            if (stream == null)
+                throw new ArgumentNullException("stream");
 
             if (typeof(IEnumerable).IsAssignableFrom(type))
                 value = (value as IEnumerable<object>).ToList();
 
-            var packer = new CompiledPacker(packPrivateField: false);
-            byte[] buffer = packer.Pack(value);
-            writeStream.Write(buffer, 0, buffer.Length);
+            var tcs = new TaskCompletionSource<object>();
+
+            try
+            {
+                var packer = SerializationContext.Default.GetSerializer(type);
+                packer.Pack(stream, value);
+
+                tcs.SetResult(null);
+            }
+            catch (Exception ex)
+            {
+                tcs.SetException(ex);
+            }
+
+            return tcs.Task;
```

Comment:
Wouldn't the result be the same using a synchronous approach but without having to use the `TaskCompletionSource`?